libraries/chain/CMakeLists.txt

```diff
@@ -129,7 +130,7 @@ add_library( eosio_chain
              )
 
 target_link_libraries( eosio_chain PUBLIC bn256 fc chainbase eosio_rapidjson Logging IR WAST WASM Runtime
-                       softfloat builtins ${CHAIN_EOSVM_LIBRARIES} ${LLVM_LIBS} ${CHAIN_RT_LINKAGE}
+                       softfloat builtins custom_appbase ${CHAIN_EOSVM_LIBRARIES} ${LLVM_LIBS} ${CHAIN_RT_LINKAGE}
```

Comment:
We have tried to keep chain from depending on appbase. It looks like this is only used for EOS_ASSERTS. I would recommend reverting this and not make chain depend on appbase.

Reply:
As now read-only trxs can run in both read and write window, we need some way to know which window we are in in OC code, where we will only evict cache entries in write window in which only one thread is executing transactions. Currently we us is_main_thread. I am going to make that changes in a separate PR.

Any other way we can get window infor from producer plugin? Passing window in push transaction? Do you think that incurs some performance penality?

Reply:
Can you use `chainbase._read_only_mode` or tie it to `chain.set_db_read_only_mode()` maybe renaming to `chain.set_read_only_mode()`?

Reply:
Thanks. I will look into that direction. 

I will revert appbase change. For the assert in `code_block_num_last_used`, we don't need it here as in `eosio_contract.cpp'`s  `apply_eosio_setcode` where `code_block_num_last_used` is called, we already have an assert it cannot be called by any read-only transactions.